src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLexer.g4

```diff
@@ -0,0 +1,438 @@
+    STRING_LITERAL,NUM_FLOAT,NUM_LONG,NUM_DOUBLE,WS,SINGLE_LINE_COMMENT,
+    BLOCK_COMMENT_BEGIN,ESC,HEX_DIGIT,VOCAB,EXPONENT,FLOAT_SUFFIX,
+
+    // Please add new tokens only in the end of list! Otherwise you break compatibility!
+    // Token for Java 1.4 language enhancements
+    ASSERT,
+
+    // Tokens for Java 1.5 language enhancements
+    STATIC_IMPORT, ENUM, ENUM_DEF, ENUM_CONSTANT_DEF, FOR_EACH_CLAUSE,
+    ANNOTATION_DEF, ANNOTATIONS, ANNOTATION, ANNOTATION_MEMBER_VALUE_PAIR, ANNOTATION_FIELD_DEF,
+    ANNOTATION_ARRAY_INIT, TYPE_ARGUMENTS, TYPE_ARGUMENT, TYPE_PARAMETERS,
+    TYPE_PARAMETER, WILDCARD_TYPE, TYPE_UPPER_BOUNDS, TYPE_LOWER_BOUNDS, AT, ELLIPSIS,
+    GENERIC_START, GENERIC_END, TYPE_EXTENSION_AND,
+
+    // Please add new tokens only in the end of list! Otherwise you break compatibility!
+
+    // token which was not included to grammar initially
+    // we need to put it to the end to maintain binary compatibility
+    // with previous versions
+    DO_WHILE,
+
+    // Tokens for Java 1.7 language enhancements
+    RESOURCE_SPECIFICATION, RESOURCES, RESOURCE,
+
+    // Tokens for Java 1.8
+    DOUBLE_COLON,  METHOD_REF, LAMBDA,
+
+    // Support of java comments has been extended
+    BLOCK_COMMENT_END,COMMENT_CONTENT,
+
+    // Need to add these here to preserve order of tokens
+    SINGLE_LINE_COMMENT_CONTENT, BLOCK_COMMENT_CONTENT, STD_ESC,
+    BINARY_DIGIT, ID_START, ID_PART, INT_LITERAL, LONG_LITERAL,
+    FLOAT_LITERAL, DOUBLE_LITERAL, HEX_FLOAT_LITERAL, HEX_DOUBLE_LITERAL,
+    SIGNED_INTEGER, BINARY_EXPONENT,
+
+    PATTERN_VARIABLE_DEF, RECORD_DEF, LITERAL_RECORD,
+    RECORD_COMPONENTS, RECORD_COMPONENT_DEF, COMPACT_CTOR_DEF,
+    TEXT_BLOCK_LITERAL_BEGIN, TEXT_BLOCK_CONTENT, TEXT_BLOCK_LITERAL_END,
+    LITERAL_YIELD, SWITCH_RULE,
+
+    LITERAL_NON_SEALED, LITERAL_SEALED, LITERAL_PERMITS,
+    PERMITS_CLAUSE
+}
+
+@header {
+import com.puppycrawl.tools.checkstyle.grammar.CommentListener;
+import com.puppycrawl.tools.checkstyle.grammar.CrAwareLexerSimulator;
+}
+
+@lexer::members {
+    /**
+     * We need to create a different constructor in order to use our
+     * own implementation of the LexerATNSimulator. This is the
+     * reason for the unused 'crAwareConstructor' argument.
+     *
+     * @param input the character stream to tokenize
+     * @param crAwareConstructor dummy parameter
+     */
+    public JavaLexer(CharStream input, boolean crAwareConstructor) {
+      super(input);
+      _interp = new CrAwareLexerSimulator(this, _ATN, _decisionToDFA, _sharedContextCache);
+    }
+
+    private CommentListener commentListener = null;
+
+    /**
+     * Sets the CommentListener for the lexer.
+     *
+     * @param commentListener the commentListener to use in this lexer
+     */
+    public void setCommentListener(CommentListener commentListener){
+            this.commentListener = commentListener;
+    }
+
+    /** Tracks the starting line of a block comment. */
+    int startLine = -1;
```

Comment:
ANTLR4 does not allow local variables in lexer rules, so these had to be global.